### PR TITLE
Oop

### PR DESCRIPTION
**Pull request by klimisa**
Converted to F#,

- TrackingId
- HandlingType
- Location
- Unlocode

### CODE REVIEW

File: Domain.FSharp/Domain.FSharp.fsproj [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (moodmosaic) [inline Domain.FSharp/Domain.FSharp.fsproj:4]: Shouldn't this be [`netstandard2.1`](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) now that there [an LTS version](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-1/) out?

```suggestion
        <TargetFramework>netstandard2.1</TargetFramework>
```

File: Domain.FSharp/Shipping/Cargo.fs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+namespace FSharp.Domain.Shipping.Cargo
+
+open System
+open System.Text.RegularExpressions
```
Comment (moodmosaic) [inline Domain.FSharp/Shipping/Cargo.fs]: I can't see where Regex is used(?)
```suggestion

```

File: Domain.FSharp/Shipping/Cargo.fs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+type TrackingId(value: Guid) =
```
Comment (moodmosaic) [inline Domain.FSharp/Shipping/Cargo.fs:9]: Style nit: In F# everything is a tuple, applied as a function argument. Even for type declarations, adding a space can make the overall style more consistent (see similar style nits below).

```suggestion
type TrackingId (value: Guid) =
```

File: Domain.FSharp/Shipping/Cargo.fs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    member this.Value = value
```
Comment (moodmosaic) [inline Domain.FSharp/Shipping/Cargo.fs:10]: Use `_` if you're not using `Value` inside `TrackingId`.

```suggestion
    member _.Value = value
```

File: Domain.FSharp/Shipping/Location.fs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+namespace FSharp.Domain.Shipping.Location
+
+open System
+open System.Text.RegularExpressions
+
+type UnLocode(value: string) =
```
Comment (moodmosaic) [inline Domain.FSharp/Shipping/Location.fs:6]:
```suggestion
type UnLocode (value : string) =
```

File: Domain.FSharp/Shipping/Location.fs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    do
+        if isNull value then raise (ArgumentNullException("value"))
```
Comment (moodmosaic) [inline Domain.FSharp/Shipping/Location.fs]:
```suggestion
        if isNull value then raise <| ArgumentNullException "value"
```

File: Domain.FSharp/Shipping/Location.fs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+    do
+        let pattern = Regex "[a-zA-Z]{2}[a-zA-Z2-9]{3}"
+        let matches = pattern.Match value
+        if not matches.Success then raise (ArgumentException("Provided value is not a valid UnLocode", "value"))
+
+    member val Value = value
+
+type Location(unlocode: UnLocode, name: string) =
```
Comment (moodmosaic) [inline Domain.FSharp/Shipping/Location.fs:18]:
```suggestion
type Location (unlocode : UnLocode, name : string) =
```

File: Domain.FSharp/Shipping/Location.fs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+    do
+        if String.IsNullOrWhiteSpace(name) then raise (ArgumentException("Provided name is not valid", "name"))
```
Comment (moodmosaic) [inline Domain.FSharp/Shipping/Location.fs]:
```suggestion
        if String.IsNullOrWhiteSpace(name) then raise (ArgumentException ("Provided name is not valid", "name"))
```

Style nit: As commented earlier.

FWIW I've only seen this style documented [here](https://blog.ploeh.dk/2018/01/29/argument-list-isomorphisms/#97ea8f87723f4b1bb995b0824b1caba4) (make sure to read the whole section). See also this style in action elsewhere, e.g. https://github.com/hedgehogqa/fsharp-hedgehog/blob/master/src/Hedgehog/Seed.fs

File: Domain.FSharp/scratchpad.fsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    member x.Value = value
+
+let m = TrackingId(Guid.NewGuid())
+
+m.Value
+
+type UnLocode(value: string) =
+
+    do
+        if value = null then raise (ArgumentNullException("value"))
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx:19]: https://latkin.org/blog/2015/05/18/null-checking-considerations-in-f-its-harder-than-you-think/

File: Domain.FSharp/scratchpad.fsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
+#r "netstandard"
+
+open System
+open System.Text.RegularExpressions
+
+type TrackingId(value: Guid) =
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx:9]:
```suggestion
type TrackingId (value : Guid) =
```

File: Domain.FSharp/scratchpad.fsx [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    member x.Value = value
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx:10]:
```suggestion
    member _.Value = value
```

File: Domain.FSharp/scratchpad.fsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+let m = TrackingId(Guid.NewGuid())
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx]:
```suggestion
let m = TrackingId <| Guid.NewGuid ()
```

File: Domain.FSharp/scratchpad.fsx [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+m.Value
+
+type UnLocode(value: string) =
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx:16]:
```suggestion
type UnLocode (value : string) =
```

File: Domain.FSharp/scratchpad.fsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+    do
+        if value = null then raise (ArgumentNullException("value"))
+
+
+    do
+        let pattern = Regex "[a-zA-Z]{2}[a-zA-Z2-9]{3}"
+        let matches = pattern.Match value
+        if not matches.Success then raise (ArgumentException("Provided value is not a valid UnLocode", "value"))
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx:25]:
```suggestion
        if not matches.Success then raise (ArgumentException ("Provided value is not a valid UnLocode", "value"))
```

File: Domain.FSharp/scratchpad.fsx [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+    member val Value = value
+
+type Location(unlocode: UnLocode, name: string) =
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx:29]:
```suggestion
type Location (unlocode : UnLocode, name : string) =
```

File: Domain.FSharp/scratchpad.fsx [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    do
+        if String.IsNullOrWhiteSpace(name) then raise (ArgumentException("Provided name is not valid", "name"))
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx:32]:
```suggestion
        if String.IsNullOrWhiteSpace(name) then raise (ArgumentException ("Provided name is not valid", "name"))
```

File: Domain.FSharp/scratchpad.fsx [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+    member val UnLocode = unlocode
+    member val Name = name
+    
+let unlocode = "xa234"
+let pattern = Regex "[a-zA-Z]{2}[a-zA-Z2-9]{3}"
+let matches = (pattern.Match unlocode).Success
+
+let u = UnLocode(unlocode)
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx]:
```suggestion
let u = UnLocode unlocode
```

File: Domain.FSharp/scratchpad.fsx [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+let l = Location(u, "Myname")
```
Comment (moodmosaic) [inline Domain.FSharp/scratchpad.fsx:43]:
```suggestion
let l = Location (u, "Myname")
```

File: Domain.FSharp/Shipping/Location.fs (hunk 8 above)
Comment (moodmosaic) [inline Domain.FSharp/Shipping/Location.fs]: Also, on `IsNullOrWhiteSpace`, see https://blog.ploeh.dk/2014/11/18/the-isnullorwhitespace-trap/.

**Review by moodmosaic — COMMENTED** [review]
I left some comments, mostly style nits, though some of them (e.g. the [The IsNullOrWhiteSpace trap](https://blog.ploeh.dk/2014/11/18/the-isnullorwhitespace-trap/)) may also be applied on the C# codebase.

File: Domain.FSharp/Shipping/Cargo.fs (hunk 4 above)
Comment (klimisa) [reply Domain.FSharp/Shipping/Cargo.fs:10]: I'll stick with conventions used by the community. It's a valid comment.

File: Domain.FSharp/Shipping/Cargo.fs (hunk 3 above)
Comment (klimisa) [reply Domain.FSharp/Shipping/Cargo.fs:9]: That's what styling Rider gives me. I'll try to change it, but I'm not promising anything.

File: Domain.FSharp/Domain.FSharp.fsproj (hunk 1 above)
Comment (klimisa) [reply Domain.FSharp/Domain.FSharp.fsproj:4]: The rest of the project is not compatible with netstandard 2.1. I will convert it to 2.1 when I'll finish the port.

File: Domain/Shipping/Cargo/Itinerary.cs [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+using Domain.Shipping.Location;
+using Domain.Shipping.Voyage;
+
+namespace Domain.Shipping.Cargo
+{
+    public class Itinerary 
+    {
+        public IReadOnlyCollection<Leg> Legs { get; }
+
+        public Itinerary(IList<Leg> legs)
```
Comment (moodmosaic) [inline Domain/Shipping/Cargo/Itinerary.cs:14]:
```suggestion
        public Itinerary(IReadOnlyCollection<Leg> legs)
```

File: Domain/Shipping/Cargo/Itinerary.cs [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            if (legs == null)
+                throw new ArgumentNullException(nameof(legs));
+            if (legs.Count == 0)
+                throw new ArgumentException("legs cannot be empty");
+
+            Legs = new ReadOnlyCollection<Leg>(legs);
+        }
+
+        public UnLocode FirstLoadLocation => Legs.First().LoadLocation;
```
Comment (moodmosaic) [inline Domain/Shipping/Cargo/Itinerary.cs:24]: If `Legs` is empty then what?

File: Domain/Shipping/Cargo/Itinerary.cs [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+
+        public DateTime FinalArrivalDate => Legs.Last().UnloadTime;
+
+        public Leg NextOf(UnLocode location)
+        {
+            var next = (Leg)null;
+            var currentFound = false;
+
+            using (var en = Legs.GetEnumerator())
+                while (en.MoveNext())
```
Comment (moodmosaic) [inline Domain/Shipping/Cargo/Itinerary.cs:38]: The extra indentation [shouldn't be necessary with C# 8](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#using-declarations).

File: Domain.FSharp/Shipping/Location.fs (hunk 8 above)
Comment (klimisa) [reply Domain.FSharp/Shipping/Location.fs]: I read the article and I agree. Let's address that when we will introduce more functional concepts.

File: Domain.FSharp/scratchpad.fsx (hunk 9 above)
Comment (klimisa) [reply Domain.FSharp/scratchpad.fsx:19]: It's scratchpad. Please do look at that file, it's only for the interactive then I copy the code to fs files

File: Domain/Shipping/Cargo/Itinerary.cs [hunk 22]
```diff
@@ -23,7 +23,7 @@ public Itinerary(IList<Leg> legs)
 
         public UnLocode FirstLoadLocation => Legs.First().LoadLocation;
 
-        public VoyageNumber FirstYoyage => Legs.First().Voyage;
```
Comment (moodmosaic) [inline Domain/Shipping/Cargo/Itinerary.cs]: *Yoy* age (age between 18-25) :stuck_out_tongue_closed_eyes:

File: Domain/Shipping/Cargo/Itinerary.cs (hunk 20 above)
Comment (klimisa) [reply Domain/Shipping/Cargo/Itinerary.cs:24]: Legs are not going to be empty it's an invariant checking that on the constructor a few lines above the statement.

File: Domain/Shipping/Cargo/Itinerary.cs (hunk 19 above)
Comment (klimisa) [reply Domain/Shipping/Cargo/Itinerary.cs:14]: Maybe I'm not sure. What IReadOnlyCollections allow? You can enlight me on that as I'm not so familiar with lists.

Comment (moodmosaic) [reply Domain/Shipping/Cargo/Itinerary.cs:14]: Or `IReadOnlyList<Leg>` perhaps. Just to be explicit that we're injecting a list and we have no plans to write to it (read only).

File: Domain/Shipping/Cargo/Itinerary.cs (hunk 20 above)
Comment (moodmosaic) [reply Domain/Shipping/Cargo/Itinerary.cs:24]: We could try declaring them (later on, in F#) as `NonEmptyList<Leg>`.

File: Domain/Shipping/Cargo/Itinerary.cs (hunk 19 above)
Comment (klimisa) [reply Domain/Shipping/Cargo/Itinerary.cs:14]: IReadOnlyList<T> accepts IList<T> from the client?

File: Domain/Shipping/Cargo/Itinerary.cs (hunk 22 above)
Comment (klimisa) [reply Domain/Shipping/Cargo/Itinerary.cs]: Legs are not going to be empty, comment [here](https://github.com/klimisa/CargoTracker2FSharp/pull/1#discussion_r365144009)

**Comment by moodmosaic** [comment]
:tada:

**Pull request merged**